RxCocoa/Common/DelegateProxyFactory.swift

```diff
@@ -0,0 +1,66 @@
+//
+//  DelegateProxyFactory.swift
+//  RxCocoa
+//
+//  Created by tarunon on 2017/06/17.
+//  Copyright © 2017 Krunoslav Zaher. All rights reserved.
+//
+
+#if !os(Linux)
+    
+#if !RX_NO_MODULE
+    import RxSwift
+#endif
+
+/**
+Define `DelegateProxy.init` for a specific object type.
+For example, in RxScrollViewDelegateProxy
+
+
+    class RxScrollViewDelegateProxy: DelegateProxy {
+        static var delegateProxyFactory = DelegateProxyFactory { (parentObject: UIScrollView) in
+            RxScrollViewDelegateProxy(parentObject: parentObject)
+        }
+    ...
+
+
+If need to extend them, chain `extended` after DelegateProxyFactory.init
+
+    class RxScrollViewDelegateProxy: DelegateProxy {
+        static var delegateProxyFactory = DelegateProxyFactory { (parentObject: UIScrollView) in
+                RxScrollViewDelegateProxy(parentObject: parentObject)
+            }
+            .extended { (parentObject: UITableView) in
+                RxTableViewDelegateProxy(parentObject: parentObject)
+            }
+    ...
+ 
+ 
+ */
+public class DelegateProxyFactory {
+    var factories: [ObjectIdentifier: ((AnyObject) -> AnyObject)]
+    public init<Object: AnyObject>(factory: @escaping (Object) -> AnyObject) {
+        factories = [ObjectIdentifier(Object.self): { factory(castOrFatalError($0)) }]
+    }
+    
+    public func extended<Object: AnyObject>(factory: @escaping (Object) -> AnyObject) -> DelegateProxyFactory {
+        guard factories[ObjectIdentifier(Object.self)] == nil else {
+            rxFatalError("The factory of \(Object.self) is duplicated. DelegateProxy is not allowed of duplicated base object type.")
+        }
+        factories[ObjectIdentifier(Object.self)] = { factory(castOrFatalError($0)) }
+        return self
+    }
+    
+    func createProxy(for object: AnyObject) -> AnyObject {
```

Comment:
We need to add unit tests for this method.

I saw that you've added `test_DelegateProxyHasNoSpecificFactory`. 

I would like us to cover this a bit better using unit tests.

```
ClassSometimesProxy1:ClassSometimesProxy2:Class1: Class2
InitialProxy1: InitialProxy2
```

```
test1
calling `createProxy ` for object of class1 -> returns initial proxy1 type

test2
calling `createProxy` for object of class 2 -> returns initial proxy2 type

test3
calling `createProxy` for object of ClassSometimesProxy2 -> return initial proxy1 type
calling `createProxy` for object of ClassSometimesProxy1 -> return initial proxy1 type
then call extend for `ClassSometimesProxy2`
calling `createProxy` for object of ClassSometimesProxy2 -> return of type ClassSometimesProxy2
calling `createProxy` for object of ClassSometimesProxy1 -> return of type ClassSometimesProxy2

test4
calling `createProxy` for object of ClassSometimesProxy2 -> return initial proxy1 type
calling `createProxy` for object of ClassSometimesProxy1 -> return initial proxy1 type
then call extend for `ClassSometimesProxy1`
calling `createProxy` for object of ClassSometimesProxy2 -> return initial proxy1 type
calling `createProxy` for object of ClassSometimesProxy1 -> return of type ClassSometimesProxy1
```